### PR TITLE
Reader | Adding accordions to Annotation menu

### PR DESCRIPTION
**Pull request by mmlumba**
This adds accordions to our annotation menu on Reader in order to let users view only the information they need to view.
accordion-annotation-menu

Connects #2161

### CODE REVIEW

File: app/assets/stylesheets/_reviewer.scss [hunk 1]
```diff
@@ -115,8 +118,9 @@ svg[aria-hidden=true],
 }
 
 .cf-comment-wrapper {
-  padding: 0 30px;
-  width: 380px;
+  padding: 0 30px 0 0;
+  height: 76vh;
```
Comment (mdbenjam) [inline app/assets/stylesheets/_reviewer.scss]: I don't think this will work. The percentages will not be constant since we have a lot of fixed height things.

File: app/assets/stylesheets/_reviewer.scss [hunk 2]
```diff
@@ -115,8 +118,9 @@ svg[aria-hidden=true],
...
+  width: 350px;
```
Comment (mdbenjam) [inline app/assets/stylesheets/_reviewer.scss]: why the change here?

File: client/app/components/Accordion.jsx [hunk 3]
```diff
@@ -17,25 +13,45 @@ const CLASS_NAME_MAPPING = {
...
+    return <Collapse accordion={accordion} className={classnames(classNames, accordionStyle)}
+      defaultActiveKey={defaultActiveKey}>
       {accordionHeaders}
     </Collapse>;
   }
 }
 
 Accordion.propTypes = {
+  accordion: PropTypes.bool,
   children: PropTypes.node,
```
Comment (mdbenjam) [inline client/app/components/Accordion.jsx]: Try this:

```
children: function (props, propName, componentName) {
    const prop = props[propName]

    let error = null
    React.Children.forEach(prop, function (child) {
      if (child.type !== AccordionHeader) {
        error = new Error('`' + componentName + '` children should be of type `AccordionHeader`.');
      }
    })
    return error
  },
```

From here: https://stackoverflow.com/questions/27366077/only-allow-children-of-a-specific-type-in-a-react-component/36424582#36424582

File: client/app/components/Accordion.jsx [hunk 4]
```diff
@@ -17,25 +13,45 @@ const CLASS_NAME_MAPPING = {
...
       style
     } = this.props;
 
     const accordionHeaders = children.map((child) => {
-      return <Panel header={child.props.title} headerClass="usa-accordion-button" key={child.props.title}>
-        <div className="usa-accordion-content">
-          {child.props.children}
-        </div>
-      </Panel>;
+      if (child.type === AccordionHeader) {
```
Comment (mdbenjam) [inline client/app/components/Accordion.jsx]: Can we remove this? And instead add a prop type that checks the child type? See below for example.

File: client/app/components/PdfSidebar.jsx [hunk 5]
```diff
@@ -143,113 +145,118 @@ export class PdfSidebar extends React.Component {
...
+                  <EditComment
+                    comment={this.props.placedButUnsavedAnnotation}
+                    id="addComment"
+                    disableOnEmpty={true}
+                    onChange={this.props.updateNewAnnotationContent}
+                    onCancelCommentEdit={this.props.stopPlacingAnnotation}
+                    onSaveCommentEdit={this.props.createAnnotation} />}
+                {comments}
+              </div>
+              <div className="cf-keyboard-shortcuts">
```
Comment (mdbenjam) [inline client/app/components/PdfSidebar.jsx]: I don't think this should be in the comments accordion.

Comment (mmlumba) [inline client/app/components/PdfSidebar.jsx]: You mean the modal or both the button and the modal?

File: app/assets/stylesheets/_reviewer.scss (hunk 2 above)
Comment (mmlumba) [inline app/assets/stylesheets/_reviewer.scss]: Looks like this with 380px:
screen shot 2017-07-10 at 5 06 21 pm

File: client/app/components/PdfSidebar.jsx (hunk 5 above)
Comment (gnakm) [inline client/app/components/PdfSidebar.jsx]: The button link for keyboard shortcut should be outside of the accordion. 

image

File: app/assets/stylesheets/_reviewer.scss (hunk 1 above)
Comment (mmlumba) [inline app/assets/stylesheets/_reviewer.scss]: Do you have any suggestions for what the height of the comment wrapper should be? This fits the first few comments (and scrolls within the comments box only) before the main scroll jumps back to the sidebar scroll.

File: app/assets/stylesheets/_reviewer.scss [hunk 6]
```diff
@@ -319,8 +323,46 @@ svg[aria-hidden=true],
   margin-top: 0;
 }
 
+[id='cf-sidebar-accordion'] {
```
Comment (mmlumba) [inline app/assets/stylesheets/_reviewer.scss]: The ID is the wrapper around the accordion (mostly used for jump to comments), whereas the class is what applies the actual styles to the accordion

Comment (mdbenjam) [inline app/assets/stylesheets/_reviewer.scss]: Confused by this, why not just provide a class?

File: client/app/components/Accordion.jsx [hunk 7]
```diff
@@ -17,25 +13,56 @@ const CLASS_NAME_MAPPING = {
...
+      return <Panel header={child.props.title} headerClass="usa-accordion-button"
+        key={child.props.title} id={child.props.id}>
+          <div className="usa-accordion-content">
+            {child.props.children}
+          </div>
+        </Panel>;
     });
 
-    return <Collapse accordion={true} className={CLASS_NAME_MAPPING[style]}>
+    const accordionStyle = classnames({
```
Comment (mdbenjam) [inline client/app/components/Accordion.jsx]: Why move this and use classnames when you cannot have multiple of classes at once?

File: client/app/components/Accordion.jsx [hunk 8]
```diff
@@ -17,25 +13,56 @@ const CLASS_NAME_MAPPING = {
 export default class Accordion extends React.PureComponent {
   render() {
     const {
+      accordion,
       children,
+      classNames,
```
Comment (mdbenjam) [inline client/app/components/Accordion.jsx]: I want to not allow the passing of classNames. This is how our button component got messed up.

File: client/app/components/PdfSidebar.jsx [hunk 9]
```diff
@@ -50,6 +52,8 @@ export class PdfSidebar extends React.Component {
     if (this.props.scrollToSidebarComment) {
       const commentListBoundingBox = this.commentListElement.getBoundingClientRect();
 
+      document.getElementById('cf-sidebar-accordion').scrollTop = 493;
```
Comment (mdbenjam) [inline client/app/components/PdfSidebar.jsx]: First off, if we want to use magic numbers, we need to put them in a constant that describes what they are.

Second off, this is a bad practice. What happens when someone leaves a couple of tags that span more than one line. Or when the document type is long and falls on a second line. This location will no longer be correct. You can get the location of the accordion header dynamically.

File: client/app/components/PdfSidebar.jsx [hunk 10]
```diff
@@ -143,112 +147,118 @@ export class PdfSidebar extends React.Component {
...
-            creatable={true}
-            options={this.generateOptionsFromTags(tagOptions)}
-            placeholder=""
-            value={this.generateOptionsFromTags(doc.tags)}
-            onChange={this.onChange}
-            selfManageValueState={true}
+        <div id="cf-sidebar-accordion">
+          <Accordion style="outline" accordion={false}
+            defaultActiveKey={['Document information', 'Categories', 'Issue tags', 'Comments']}
+            classNames={['cf-sidebar-accordion']}>
```
Comment (mdbenjam) [inline client/app/components/PdfSidebar.jsx]: Is there a way we can wrap this in a div with the necessary css, rather than passing the class name directly?

File: client/app/components/Accordion.jsx (hunk 7 above)
Comment (mmlumba) [inline client/app/components/Accordion.jsx]: For this specific accordion I needed a class for the Reader styles (`cf-sidebar-accordion`) and the class for the accordion style in the same Accordion div.

File: client/app/components/PdfSidebar.jsx (hunk 9 above)
Comment (mmlumba) [inline client/app/components/PdfSidebar.jsx]: I originally added this line so that the comments section is the only thing that appears in the sidebar when the user clicks on a comment to jump to it. If we could either have the option to have one giant scrollbar for the comments or one scrollbar for the comments wrapper then I can get rid of this line.

File: app/assets/stylesheets/_reviewer.scss [hunk 11]
```diff
@@ -114,12 +117,20 @@ svg[aria-hidden=true],
   padding: 0 30px;
 }
 
+.cf-category-sidebar {
+  overflow-y: hidden;
+}
+
+.cf-issue-tag-sidebar {
+  overflow: visible;
+  height: 255px;
```
Comment (mdbenjam) [inline app/assets/stylesheets/_reviewer.scss]: I don't think fixed height is a good idea here. Either just let it overflow or we need to make it dynamic. I think overflow is sufficient.

File: spec/feature/reader_spec.rb [hunk 12]
```diff
@@ -483,7 +483,7 @@ def placing_annotation_icon_position
 
         click_on documents[0].type
 
-        element = "cf-comment-wrapper"
+        element = "cf-sidebar-accordion"
```
Comment (mdbenjam) [inline spec/feature/reader_spec.rb]: Do we have a test for the accordion component? We should probably either add one, or add something in a feature test.

File: app/assets/stylesheets/_reviewer.scss (hunk 11 above)
Comment (mmlumba) [inline app/assets/stylesheets/_reviewer.scss]: This is what happens with just overflow:
issue-tag


Is this ok with you and @gnakm ?

Comment (gnakm) [inline app/assets/stylesheets/_reviewer.scss]: I think this is good for now. I would prefer it to be dynamic because overflow makes the feature hidden and I wonder if it's going to cause frustration for our users. Since we've had this open for a while, how about closing it and discussing making it dynamic in another issue?

File: client/app/components/Accordion.jsx [hunk 13]
```diff
@@ -17,25 +18,49 @@ const CLASS_NAME_MAPPING = {
...
-  children: PropTypes.node,
+  accordion: PropTypes.bool,
+  children (props, propName, componentName) {
+    const prop = props[propName];
+
+    let error = null;
+
+    React.Children.forEach(prop, (child) => {
+      if (child.type !== AccordionHeader) {
+        error = new Error(`\`${componentName}\` children should be of type \`AccordionHeader\`.`);
```
Comment (NickHeiner) [inline client/app/components/Accordion.jsx]: I generally like to include what the problematic value is in error message as well.

```js
new Error(`\`${componentName}\` children should be of type \`AccordionHeader\`, but was \`${child.type}\`.`);
```

File: client/app/components/PdfSidebar.jsx [hunk 14]
```diff
@@ -145,112 +147,120 @@ export class PdfSidebar extends React.Component {
...
-                buttons = {[
-                  { classNames: ['usa-button', 'usa-button-secondary'],
-                    name: 'Thanks, got it!',
-                    onClick: this.toggleKeyboardModal
+        <div className="cf-sidebar-accordion" id="cf-sidebar-accordion" ref={(commentListElement) => {
+          this.commentListElement = commentListElement;
+        }}>
+          <Accordion style="outline" accordion={false}
+            defaultActiveKey={['Document information', 'Categories', 'Issue tags', 'Comments']}>
+            <AccordionHeader title="Document information" key={1}>
```
Comment (NickHeiner) [inline client/app/components/PdfSidebar.jsx]: Let's use a descriptive `key` instead of numbering them. That way, we won't have problems when we add or remove items from the list and then have to re-number everything.

```
key='document-information'
```

File: spec/feature/reader_spec.rb [hunk 15]
```diff
@@ -650,13 +650,39 @@ def placing_annotation_icon_position
...
 
       click_on "Open menu"
       expect(page).to have_content("Document Type")
     end
 
+    scenario "Open and close accordion sidebar menu" do
+      visit "/reader/appeal/#{appeal.vacols_id}/documents/"
+      click_on documents[0].type
+
+      # Using xpath selectors to click on divs for each section
```
Comment (NickHeiner) [inline spec/feature/reader_spec.rb]: Why are we using `xpath` instead of standard CSS style selectors?

Comment (mmlumba) [inline spec/feature/reader_spec.rb]: Xpath selectors are more precise to get the right accordion header to click on as opposed to css selectors.

File: client/app/components/Accordion.jsx (hunk 13 above)
Comment (NickHeiner) [inline client/app/components/Accordion.jsx]: image

Ha, so this error message actually turns out to not look very good. Looks like putting `child.type` directly in the string does not format nicely. And because we have a newline in the template string, the error message is awkwardly broken up into two lines.

Comment (NickHeiner) [inline client/app/components/Accordion.jsx]: Also, that error message always pops up when I load the page. That is not good.

**Comment by NickHeiner** [comment]
I ran the test 20 times locally to ensure stability:

image

**Comment by NickHeiner** [comment]
Here's the [test passing 20 times on Travis](https://travis-ci.org/department-of-veterans-affairs/caseflow/builds/255744102?utm_source=github_status&utm_medium=notification):

image

File: vendor/assets/stylesheets/_rc_collapse.scss [hunk 16]
```diff
@@ -41,7 +41,8 @@ $borderStyle: 1px solid #d9d9d9;
     &:first-child {
       border-top: none;
     }
-    margin-bottom: 6px;
+    // Caseflow deviation from the original package:
+    // margin-bottom: 6px;
```
Comment (mdbenjam) [inline vendor/assets/stylesheets/_rc_collapse.scss]: Is this supposed to be commented out?

File: client/app/reader/reducer.js [hunk 17]
```diff
@@ -658,6 +661,9 @@ export const reducer = (state = initialState, action = {}) => {
         pdf: {
           isPlacingAnnotation: { $set: true }
         }
+      },
+      openedAccordionSections: {
+        $apply: (sectionKeys) => _.union(sectionKeys, [Constants.COMMENT_ACCORDION_KEY])
```
Comment (mdbenjam) [inline client/app/reader/reducer.js]: I like this. I'd love to do it for when you jump to comments from the list view as well--but that'll have to wait until we move that path to redux :(

File: client/app/reader/actions.js [hunk 18]
```diff
@@ -448,3 +448,10 @@ export const addNewTag = (doc, tags) => (
     }
   }
 );
+
+export const setOpenedAccordionSections = (openedAccordionSections) => ({
```
Comment (mdbenjam) [inline client/app/reader/actions.js]: We should probably add an analytics event where this action is called.

`Analytics.event('Accordion toggled', '<accordion section>', '');`

File: vendor/assets/stylesheets/_rc_collapse.scss (hunk 16 above)
Comment (NickHeiner) [inline vendor/assets/stylesheets/_rc_collapse.scss]: Yes. I was hoping that the comment above would make that obvious haha. :smile:

I feel weird about the way we've copied this third party package into our repo instead of using a package manager. (I wasn't in the conversation where this was decided, so maybe I'm missing something.) When I'm in the "someone else's file that I have to copy-paste here" mode, I tend to make changes like this: I clearly mark where I've made alterations, and prefer to just comment things out rather than remove them entirely.

I can also just yank the line if you'd prefer.

File: client/app/reader/constants.js [hunk 19]
```diff
@@ -68,6 +68,10 @@ export const SET_OPENED_ACCORDION_SECTIONS = 'SET_OPENED_ACCORDION_SECTIONS';
 
 export const COMMENT_ACCORDION_KEY = 'Comments';
 
+export const ANALYTICS = {
+  VIEW_DOCUMENT_PAGE: 'Document Viewer'
```
Comment (mdbenjam) [inline client/app/reader/constants.js]: ooo a constant is a good idea.

**Comment by mdbenjam** [comment]
I was just playing around with it some more and realized that there is a horizontal scroll bar in the comment section. I think it'd be good to remove that. Seems like these two rules should do the trick.
screen shot 2017-07-20 at 5 30 56 pm

**Comment by mdbenjam** [comment]
diff --git a/app/assets/stylesheets/_reviewer.scss b/app/assets/stylesheets/_reviewer.scss
index a8f8b02f2..f711caf6a 100644
--- a/app/assets/stylesheets/_reviewer.scss
+++ b/app/assets/stylesheets/_reviewer.scss
@@ -140,8 +140,6 @@ svg[aria-hidden=true],
 }
 
 .cf-comment-wrapper {
-  padding: 0 30px 0 0;
-  width: 350px;
   $comment-wrapper-vertical-margin: 15px;
   margin-top: $comment-wrapper-vertical-margin;
